### PR TITLE
Fix bugs

### PR DESCRIPTION
**Pull request by MartaAG**
For evaluation.

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -16,7 +16,7 @@
           </v-container>
         </v-main>
         <v-footer color="brown darken-4" app dark>
-          <span>2020</span>
+          <span class="white--text">&copy; {{ new Date().getFullYear() }}</span>
```
Comment (Udiknedormin) [inline src/App.vue:19]: Re-introduced the local date issue here.

File: src/App.vue [hunk 2]
```diff
@@ -1,40 +1,26 @@
...
+              <v-col cols="12">
+                <router-view></router-view>
+              </v-col>
+            </v-row>
+          </v-container>
+        </v-main>
+        <v-footer color="brown darken-4" app dark>
+          <span class="white--text">&copy; {{ new Date().getFullYear() }}</span>
+        </v-footer>
+      </v-app>
```
Comment (Udiknedormin) [inline src/App.vue:21]: Formatting-wise --- much better. Re-introducing-a-bug-wise --- not so cool

**Comment by Udiknedormin** [comment]
```
$npm install
(...)
found 1 high severity vulnerability
(...)
```
^^^
please fix it!

File: src/components/RestaurantDetails.vue [hunk 3]
```diff
@@ -29,48 +36,45 @@
       <h4>Linki</h4>
       <ul>
         <li>
-          <a :href="restaurant.url">{{restaurant.name}}</a><br/>
-          <a :href="restaurant.photos_url">Photos</a><br/>
-          <a :href="restaurant.menu_url">Menu</a><br/>
-          <a :href="restaurant.events_url">Events</a><br/>
+          <a :href="restaurant.url">{{ restaurant.name }}</a
+          ><br />
```
Comment (Udiknedormin) [inline src/components/RestaurantDetails.vue:40]: Breaking a line line that (i.e. in the middle of a short tag closing) is one of the worse style choices one can make.

File: src/main.js [hunk 4]
```diff
@@ -6,11 +6,31 @@ import vuetify from "./plugins/vuetify";
...
+axios.interceptors.response.use(
+  response => {
+    if (response.status === 403) {
+      console.log(response);
+    }
+    return response;
+  },
+  error => {
+    if (error.response.status === 403) {
+      alert("please use your API key to run the app");
```
Comment (Udiknedormin) [inline src/main.js:25]: What if I did and yet got 403? Tokens and keys can expire. Not to mention you're not being exactly nice to the user --- do you really expect they read the WHOLE readme before using your site? Well, now they need to!

Note that `alert` windows can be quite big and still remain readable... and you can copy-paste from them!

File: src/store/index.js [hunk 5]
```diff
@@ -6,21 +6,34 @@ Vue.use(Vuex);
...
+        const response = await axios.get("/search", {
+          params: {
+            entity_id: 263,
+            entity_type: "city",
+            sort: "rating"
+          }
+        });
 
-      commit("setRestaurants", response.data.restaurants);
+        commit("setRestaurants", response.data.restaurants);
```
Comment (Udiknedormin) [inline src/store/index.js:33]: Modifying the list in-place has both advantages and disadvantages.

Pros:
* slightly more performant --- but these arrays are so tiny it hardly matters
Cons:
* impossible to check whenever the default value was used or not afterwards --- could be useful

**Comment by Udiknedormin** [comment]
A BIG problem with your alert: let's say I did realise my api key just expired or I mis-pasted it. I change it... and the communicate stays the same. If I was a user, I'd be puzzled as for why. A smart user would probably try to rebuild the project --- and that, indeed, helps --- but you know... the user was not-so-smart enough not to read README before using your service, so I'd not count on user's smartness if I were you.

File: src/App.vue (hunk 1 above)
Comment (Udiknedormin) [reply src/App.vue:19]: Update:

Here, is I change my date system settings to, say, 2019, the site will show copyright 2019. But that is false, as your site wasn't even there in 2019. You spread a false information then, and false copyright information is a serious legal issue.

File: src/components/RestaurantDetails.vue [hunk 6]
```diff
@@ -36,8 +36,7 @@
       <h4>Linki</h4>
       <ul>
         <li>
-          <a :href="restaurant.url">{{ restaurant.name }}</a
-          ><br />
+          <a :href="restaurant.url">{{ restaurant.name }}</a><br />
```
Comment (Udiknedormin) [inline src/components/RestaurantDetails.vue]: Please make separate commits for separate issues, especially if they're in separate files (it's not a problem to first commit one file and then the other one).

Some programmers even claim that style changes should have their own PRs but I'm not as big of a fundamentalist.

File: src/main.js [hunk 7]
```diff
@@ -11,18 +11,22 @@ Vue.use(VueAxios, axios);
...
     return response;
   },
   error => {
-    if (error.response.status === 403) {
-      alert("please use your API key to run the app");
+    if (!process.env.VUE_APP_API_KEY) {
+      alert(
+        "Please make a file named '.env.local'. Create a global variable VUE_APP_API_KEY and initialize it with your own key in this file. " +
+          error
+      );
```
Comment (Udiknedormin) [inline src/main.js:26]: Much better, the solution is almost what I'd expect, there are very few "buts":
* "create a global variable", "initialize" --- what if I'm not a technical person? I understand "create/make a file", but what is "create a global variable (...) and initialize it (...) in your file"?
* I followed the instructions, closed the window... and nothing happens. Reloaded. Nothing happens. If I wasn't a technical person, it's not a given I'd think of restarting the server just to update the key.
* you could provide a link to documentation, NEVER be afraid of it
* you don't SAY what the actual error was, just provide a solution and the "cause" error 403 --- I, as a user, don't really know why I should do that
* are you sure showing the "cause" error is needed at all here? The server did respond, so the network connection is fine and 403 (Forbidden) is what you get on authorization fails --- but you do know the cause of the authorization failure. The other alert (a few lines below) is much better, it uses the fact you already know the key exists, so if it still yielded 403 then the token may be invalid or expired, but certainly not absent.

All of that being said, `alert` isn't a really nice way to show errors like that, as it "breaks the wall" of your app. E.g. if your app is all bluish-greenish, and the alert window is black/dark-gray (in the browser I'm using to type this comment --- it is), then it feels off and as if it was the browser-related issue, not page reporting some casual page-related problem.

File: src/main.js [hunk 8]
```diff
@@ -11,18 +11,22 @@ Vue.use(VueAxios, axios);
...
+    }
+    if (process.env.VUE_APP_API_KEY && error.response.status === 403) {
```
Comment (Udiknedormin) [inline src/main.js:28]: What about other error codes? Are there any other kinds of errors you think you could handle?

File: src/main.js [hunk 9]
```diff
@@ -11,18 +11,22 @@ Vue.use(VueAxios, axios);
...
+      alert("Your key expired or is invalid  " + error);
```
Comment (Udiknedormin) [inline src/main.js:29]: Could provide a suggested solution. A link to Zomato's page with the description of how to get a key would be very appreciated by the user who got it all wrong (e.g. entered their github key instead of zomato key).

File: README.md [hunk 10]
```diff
@@ -1,5 +1,7 @@
 # best-restaurants
-Shows 20 restaurants with best ratings from Wroclaw
+Shows 20 restaurants with best ratings from Wroclaw.
+The application bases on the rating from zomatoAPI.
```
Comment (Udiknedormin) [inline README.md:3]: "ZamatoAPI", or more precisely: the fact that your way to get the Zomato data is via REST API, is an implementation detail. "Zomato" is enough.

File: src/main.js [hunk 11]
```diff
@@ -20,13 +19,25 @@ axios.interceptors.response.use(
...
   error => {
     if (!process.env.VUE_APP_API_KEY) {
-      alert(
-        "Please make a file named '.env.local'. Create a global variable VUE_APP_API_KEY and initialize it with your own key in this file. " +
-          error
-      );
+      console.log(error);
+      router.replace({
+        name: "Error",
+        params: { id: 1 }
```
Comment (Udiknedormin) [inline src/main.js:25]: That `1` here looks like a "magic number". It would be better to use some named constant.

File: src/main.js [hunk 12]
```diff
@@ -20,13 +19,25 @@ axios.interceptors.response.use(
...
+        name: "Error",
+        params: { id: error.response.status }
+      });
+    }
+    if (process.env.VUE_APP_API_KEY && error.response.status === 404) {
+      console.log(error);
+      router.replace({
+        name: "Error",
+        params: { id: error.response.status }
+      });
```
Comment (Udiknedormin) [inline src/main.js:40]: The only difference between the bodies of these three `if`s is that in the first one, status is replaced by a constant `1`. Seems like it could be factored out or at least simplified.

File: src/router/index.js [hunk 13]
```diff
@@ -18,6 +19,11 @@ const routes = [
     // which is lazy-loaded when the route is visited.
     component: () =>
       import(/* webpackChunkName: "about" */ "../views/About.vue")
+  },
+  {
+    path: "/error/:id",
+    name: "Error",
+    component: DisplayError
```
Comment (Udiknedormin) [inline src/router/index.js:26]: `DisplayError`, "display error", is a imperative verb expression. Component names should be nouns though.

File: src/views/DisplayError.vue [hunk 14]
```diff
@@ -0,0 +1,37 @@
+<template>
+  <div>
```
Comment (Udiknedormin) [inline src/views/DisplayError.vue:2]: Why a `div` here? Its contents is a single tag, so it's not a must-have in this case.

File: src/views/DisplayError.vue [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+</template>
+
+<script>
+export default {
+  name: "DisplayError",
+  computed: {
+    errorMessage: function() {
+      let message = "";
+      switch (this.$route.params.id) {
+        case 1:
```
Comment (Udiknedormin) [inline src/views/DisplayError.vue:19]: As before --- `1` is a "magic number" here. It doesn't mirror any HTTP code (unlike 403 and 404), just a "random" value. It would be fine if it was named, but seeing it in the caller code looks like this `1` has some meaning (possibly "the first one" or "some positive integer").

File: src/views/DisplayError.vue [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+          break;
+        case 403:
+          message =
+            "Your key expired or is invalid. Please visit https://developers.zomato.com/api#headline2 and request a new API key";
+          break;
+        case 404:
+          message = "Page you requested does not exist";
+          break;
+        default:
+          message = "Unknown error";
```
Comment (Udiknedormin) [inline src/views/DisplayError.vue:31]: Please run your local server and manually go to `localhost:8080/error/404`. What is the message? Can you guess why?

File: src/views/DisplayError.vue [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+<script>
+export default {
+  name: "DisplayError",
+  computed: {
+    errorMessage: function() {
+      let message = "";
+      switch (this.$route.params.id) {
+        case 1:
+          message =
+            "Your API key is absent. To run the app, please make a file named  '.env.local'. Please visit https://developers.zomato.com/api#headline2 and request your API key. Then write in env.local: VUE_APP_API_KEY=your API KEY. Save it. Run in console npm run serve. For more information, read readme file.";
```
Comment (Udiknedormin) [inline src/views/DisplayError.vue:21]: "Please" isn't needed, as it's not really a request. It's instruction. Other than that, "Then write in env.local" starts with a capital letter after comma and lacks the dot prefix in filename.

Also: it would be even better if there was a link instead of an URL (and also a link to the readme, while we're at that).

File: src/views/DisplayError.vue [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+          break;
+        case 403:
+          message =
+            "Your key expired or is invalid. Please visit https://developers.zomato.com/api#headline2 and request a new API key";
+          break;
+        case 404:
+          message = "Page you requested does not exist";
```
Comment (Udiknedormin) [inline src/views/DisplayError.vue:28]: I, as a user, requested `localhost:8080`. I dare to say that it DOES exist...

**Review by Udiknedormin — CHANGES_REQUESTED** [review]

File: src/views/DisplayError.vue (hunk 17 above)
Comment (MartaAG) [reply src/views/DisplayError.vue:21]: > key. Then 

Here? There is no comma here?

**Comment by Udiknedormin** [comment]
```
$ npm install
$ npm run serve
 WARNING  Compiled with 1 warnings                                      18:37:16

Module Warning (from ./node_modules/eslint-loader/index.js):

$MartaAG_bestRestaurants/src/views/ErrorCodes.vue
  24:14  warning  Replace `'403'` with `"403"`  prettier/prettier
  29:14  warning  Replace `'404'` with `"404"`  prettier/prettier

✖ 2 problems (0 errors, 2 warnings)
  0 errors and 2 warnings potentially fixable with the `--fix` option.

(...)
```
^ please fix it

File: src/views/ErrorCodes.vue [hunk 19]
```diff
@@ -0,0 +1,39 @@
+<template>
+  <v-alert type="error" color="pink darken-4">
+    <v-row align="center">
+      <v-col class="grow">
+        <div v-html="errorMessage"></div>
+      </v-col>
+      <v-col class="shrink"> </v-col>
+    </v-row>
+  </v-alert>
+</template>
```
Comment (Udiknedormin) [inline src/views/ErrorCodes.vue:10]: How can I go back to the main page once I do what the error message tells me to do? On "about" pages I have the "Back" button, but I can't see any analogue here.

**Review by Udiknedormin — CHANGES_REQUESTED** [review]
Buggy!

File: src/views/ErrorCodes.vue [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+              Your key expired or is invalid. Visit <a target="_blank" rel="noopener noreferrer" class="text-decoration-none font-weight-black pink--text text--lighten-3" href="https://developers.zomato.com/api#headline2">Zomato</a> and request a new API key. For more information, read <a target="_blank" rel="noopener noreferrer" href="https://github.com/MartaAG/bestRestaurants/blob/fixBugs/README.md" class="text-decoration-none font-weight-black pink--text text--lighten-3">Readme</a> file.
+            </p>`;
+          break;
+        case '404':
+          message = "Page you requested does not exist";
+          break;
+        default:
+          message = "Unknown error";
+      }
+      return message;
```
[the remark this answers is not in the document]
Comment (MartaAG) [reply src/views/ErrorCodes.vue:35]: Well, that's the structure of switch expression, or am I missing something?

File: src/store/index.js [hunk 21]
```diff
@@ -34,17 +39,28 @@ export default new Vuex.Store({
...
+        const response = await axios.get("/restaurant", {
+          params: {
+            res_id: ID
+          }
+        });
+
+        commit("setCurrentRestaurant", response.data);
+      } catch (error) {
+        console.log(error.message);
+      }
```
Comment (Udiknedormin) [inline src/store/index.js]: It does fix the redirection problem, via interceptor. Two problems here though:

1. I already fetched quite a big list of restaurants but not I need to re-fetch a single one each time I want to see the details? It will drain my Zomato key pretty quickly (many API keys have limits on how many queries you may perform in a unit of time), don't you think?

2. I went to "/about/16536386" and successfully viewed the details of Cafe La Ruina, a top-rated cafe in Poznań. But your service was supposed to inform me about food places in Wrocław, not Poznań, wasn't it?

**Review by Udiknedormin — CHANGES_REQUESTED** [review]
"Go Back" button for non-404 errors.

API key being drained out.

No protection from viewing food places in Poznań instead of Wrocław!

File: src/views/ErrorCodes.vue [hunk 22]
```diff
@@ -4,7 +4,9 @@
       <v-col class="grow">
         <div v-html="errorMessage"></div>
       </v-col>
-      <v-col class="shrink"> </v-col>
+      <v-col class="shrink" v-if="this.$route.params.id === '404'">
```
[the remark this answers is not in the document]
Comment (MartaAG) [reply src/views/ErrorCodes.vue:7]: For the first two alerts, I considered it, but there is nothing you can return to. There is only a piece of information that key is missing/invalid - so one will know he will get a proper key, run the server from console... and that's all.

File: src/components/RestaurantDetails.vue [hunk 23]
```diff
@@ -1,71 +1,84 @@
...
-    </v-btn>
+    <v-row>
+      <v-col lg="4">
+        <v-card class="mx-auto" max-width="500" elevation="15">
+          <v-card-text>
+            <h2 class="display-1 text--primary">{{ restaurant.name }}</h2>
+            <h3 class="text--primary">{{ restaurant.location.address }}</h3>
+            <p>Hours: {{ restaurant.timings }}</p>
+            <p>Phone: {{ restaurant.phone_numbers }}</p>
+          </v-card-text>
```
[the remark this answers is not in the document]
Comment (MartaAG) [reply src/components/RestaurantDetails.vue:11]: I thought it would be a better way to emphasize the most crucial information about the restaurant: name, address and its telephone

File: src/store/index.js (hunk 21 above)
Comment (MartaAG) [reply src/store/index.js]: I added cache to not drain the key. 
The problem is I got stuck with showing only Wroclaw's food places.
When I use 'restaurant' the only parameter you can use is ID of the restaurant.  
I have no clue what to do.

File: src/components/RestaurantDetails.vue [hunk 24]
```diff
@@ -1,71 +1,84 @@
...
+        </v-card>
+        <br />
+        <p>
+          <span class="font-weight-medium">Cuisines:</span>
+          {{ restaurant.cuisines }}
+        </p>
+        <p>
+          <span class="font-weight-medium"
+            >Average cost for two customers:</span
```
[the remark this answers is not in the document]
Comment (MartaAG) [reply src/components/RestaurantDetails.vue]: I don't know but sometimes prettier linter does this weird formatting. 
And when I correct it, linter still gives me warnings

File: src/components/RestaurantDetails.vue (hunk 23 above)
[the remark this answers is not in the document]
Comment (Udiknedormin) [reply src/components/RestaurantDetails.vue]: Hmm, I see. A matter of taste, I guess.

File: src/views/ErrorCodes.vue (hunk 22 above)
[the remark this answers is not in the document]
Comment (Udiknedormin) [reply src/views/ErrorCodes.vue:7]: Well, when I fix the API key and rebuild the server, I'm still on the error page. If I had a "back" button, I could try again... or should it be called "retry"?

File: src/store/index.js (hunk 21 above)
Comment (Udiknedormin) [reply src/store/index.js]: Hint: look at the [Zomato API swagger](https://developers.zomato.com/documentation).

File: src/store/index.js [hunk 25]
```diff
@@ -34,33 +40,36 @@ export default new Vuex.Store({
...
+          const response = await axios.get("/restaurant", {
+            params: {
+              res_id: ID
+            }
+          });
+          commit("setRestaurant", response.data);
+          commit("addToCache", response.data)
+        } catch (error) {
+          console.log(error.message);
+        }
```
Comment (Udiknedormin) [inline src/store/index.js:63]: Now, if I go to the same "about" page twice, it's cached. However, when I go from the list of best restaurants to a single details page, it will still be reloaded. So it still takes N+1 queries, where N is the number of best restaurants listed on the main page.

File: src/components/RestaurantDetails.vue [hunk 26]
```diff
@@ -2,13 +2,32 @@
...
+            text-subtitle-1">
+              <span class="font-weight-medium">Hours:</span>
+              {{ restaurant.timings }}
+            </p>
+
+            <p class="pink--text text--lighten-5
+            text-subtitle-1">
+              <span class="font-weight-medium">Phone:</span>
+              {{ restaurant.phone_numbers }}
+            </p>
```
Comment (Udiknedormin) [inline src/components/RestaurantDetails.vue:39]: Looks much better now.

**Review by Udiknedormin — CHANGES_REQUESTED** [review]
Not enough caching. Lack of "try again" button.
